From 828ac4b7981232afce12b0cfbb79189508c7765e Mon Sep 17 00:00:00 2001
From: Aswin M <aswin_m@comcast.com>
Date: Tue, 8 Nov 2022 12:21:18 +0000
Subject: [PATCH] HX44X-149 : Bring in Voice Support

Reason for change: Add support for interactive authendication screen.
Test Procedure: Enable necessary configurations and build with avs-sdk-oobe-screens.
Risks: Low
License: Inherited
Upstream-Status: Pending
Source: RDKM

Signed-off-by: Aswin M <Aswin_M@comcast.com>
---
 css/main.css      |   7 +-
 js/ss-sdk-oobe.js | 396 ++++++++++++++++++++++++++++++++++++++++++++--
 ss-sdk.html       |   6 +-
 3 files changed, 398 insertions(+), 11 deletions(-)

diff --git a/css/main.css b/css/main.css
index b6fc36d..4aa6aea 100644
--- a/css/main.css
+++ b/css/main.css
@@ -94,4 +94,9 @@ body {
 .btn-primary:active, .btn-primary:hover {
     background-color: #008cff;
     border-color: #008cff;
-}
\ No newline at end of file
+}
+
+.btn-secondary:active, .btn-secondary:hover, .btn-secondary:focus {
+    background-color: #008cff;
+    border-color: #008cff;
+}
diff --git a/js/ss-sdk-oobe.js b/js/ss-sdk-oobe.js
index 8c0579b..07fee00 100755
--- a/js/ss-sdk-oobe.js
+++ b/js/ss-sdk-oobe.js
@@ -1,3 +1,10 @@
+var HandledReqAuthMsg = 0;
+var LoginSkip = 0;
+var TabIndexCounter = 0;
+var BtnElements =[];
+var LstPrsdKey = '';
+var prvspage = '';
+
 var setupOptions = {
     cblData: null,
     currentScreen: null,
@@ -10,9 +17,12 @@ var textStrings = {
 }
 
 var onHandleRequestAuthorization = (requestAuthorizationMessage) => {
-    console.log('onHandleRequestAuthorization called!');
-    console.log(requestAuthorizationMessage);
+  console.log('onHandleRequestAuthorization called!');
+  console.log(requestAuthorizationMessage);
 
+  if(!LoginSkip)
+  {
+    HandledReqAuthMsg = 1;
     // set CBL values
     cblCodeAcquired(requestAuthorizationMessage);
     console.log(setupOptions.currentScreen);
@@ -43,13 +53,16 @@ var onHandleRequestAuthorization = (requestAuthorizationMessage) => {
             navigation.ToStart();
             break;
     }
+  }
 }
 
 var onHandleAuthorizationStateChanged = (authStateChangeMessage) => {
-    console.log('onHandleAuthorizationStateChanged called!');
-    console.log(authStateChangeMessage);
+  console.log('onHandleAuthorizationStateChanged called!');
+  console.log(authStateChangeMessage);
+  if((!LoginSkip)&&(HandledReqAuthMsg))
+  {
 
-    switch(authStateChangeMessage.state){
+    switch(authStateChangeMessage){
         case 'UNINITIALIZED':
             // error - uninitialized
             navigation.ToError('UNINITIALIZED');
@@ -67,23 +80,86 @@ var onHandleAuthorizationStateChanged = (authStateChangeMessage) => {
             navigation.ToError('ERROR');
             break;
     }
+   }
 }
 
 var navigation = {
     ToStart: () => {
         setupOptions.currentScreen = "start";
+        document.body.style.setProperty('background-color', '#1A2731', 'important');
         document.getElementById("root").style.display = "none"; // SS SDK container
         document.getElementById("ss_sdk_oobe").style.display = "block"; // OOBE container
         document.getElementById("screen_default_0").style.display = "block";
         document.getElementById("screen_default_cbl").style.display = "none";
         document.getElementById("screen_default_success").style.display = "none";
         document.getElementById("screen_default_error").style.display = "none";
-        document.getElementById("bt_back").style.display = "block";
+        document.getElementById("bt_back").style.display = "none";
         document.getElementById("bt_done").style.display = "none";
-        backAction = () => {};
+        document.getElementById("bt_skip").style.display = "block";
+		
+		ClrElementFocus('start');
+		BtnElements =[document.getElementById("bt_sign_in"),document.getElementById("bt_skip")];
+	
+		var SetVisibility = new XMLHttpRequest();
+	
+		SetVisibility.onreadystatechange = function() {
+		if (this.readyState == 4 && this.status == 200) {
+		
+			var ZorderChange = new XMLHttpRequest();
+
+			ZorderChange.onreadystatechange = function() {
+			if (this.readyState == 4 && this.status == 200) {
+		
+				var FocusReq = new XMLHttpRequest();
+				FocusReq.open("POST", "http://127.0.0.1:9998/jsonrpc");
+				FocusReq.setRequestHeader("Accept", "application/json");
+				FocusReq.setRequestHeader("Content-Type", "application/json");
+		
+				let Focuscmd = `{
+				"jsonrpc": "2.0",
+				"id": 3,
+				"method": "org.rdk.RDKShell.1.setFocus",
+				"params": { "client": "SmartScreen"}
+				}`;
+			
+				FocusReq.send(Focuscmd);
+			}
+			};
+			
+			ZorderChange.open("POST", "http://127.0.0.1:9998/jsonrpc");
+			ZorderChange.setRequestHeader("Accept", "application/json");
+			ZorderChange.setRequestHeader("Content-Type", "application/json");
+		
+			let MoveToBackcmd = `{
+			"jsonrpc": "2.0",
+			"id": 3,
+			"method": "org.rdk.RDKShell.1.moveToBack",
+			"params": { "client": "ResidentApp"}
+			}`;
+		
+			ZorderChange.send(MoveToBackcmd);			
+			
+		}
+		};
+		
+	
+		SetVisibility.open("POST", "http://127.0.0.1:9998/jsonrpc");
+		SetVisibility.setRequestHeader("Accept", "application/json");
+		SetVisibility.setRequestHeader("Content-Type", "application/json");
+	
+		let SetVisibilityCmd = `{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"method": "org.rdk.RDKShell.1.setVisibility",
+		"params": { "client": "SmartScreen" , "visible": true }
+		}`;
+
+		SetVisibility.send(SetVisibilityCmd);
+        backAction = navigation.ToDone;
     },
     ToCBL: () => {
         setupOptions.currentScreen = "cbl";
+        document.body.style.setProperty('background-color', '#1A2731', 'important');
         document.getElementById("root").style.display = "none"; // SS SDK container
         document.getElementById("ss_sdk_oobe").style.display = "block"; // OOBE container
         document.getElementById("screen_default_0").style.display = "none";
@@ -92,10 +168,74 @@ var navigation = {
         document.getElementById("screen_default_error").style.display = "none";
         document.getElementById("bt_back").style.display = "block";
         document.getElementById("bt_done").style.display = "none";
+        document.getElementById("bt_skip").style.display = "none";
+		
+		ClrElementFocus('cbl');
+
+		BtnElements =[document.getElementById("bt_back")];
+
+		var SetVisibility = new XMLHttpRequest();
+	
+		SetVisibility.onreadystatechange = function() {
+		if (this.readyState == 4 && this.status == 200) {
+		
+			var ZorderChange = new XMLHttpRequest();
+
+			ZorderChange.onreadystatechange = function() {
+			if (this.readyState == 4 && this.status == 200) {
+		
+				var FocusReq = new XMLHttpRequest();
+				FocusReq.open("POST", "http://127.0.0.1:9998/jsonrpc");
+				FocusReq.setRequestHeader("Accept", "application/json");
+				FocusReq.setRequestHeader("Content-Type", "application/json");
+		
+				let Focuscmd = `{
+				"jsonrpc": "2.0",
+				"id": 3,
+				"method": "org.rdk.RDKShell.1.setFocus",
+				"params": { "client": "SmartScreen"}
+				}`;
+			
+				FocusReq.send(Focuscmd);
+			}
+			};
+			
+			ZorderChange.open("POST", "http://127.0.0.1:9998/jsonrpc");
+			ZorderChange.setRequestHeader("Accept", "application/json");
+			ZorderChange.setRequestHeader("Content-Type", "application/json");
+		
+			let MoveToBackcmd = `{
+			"jsonrpc": "2.0",
+			"id": 3,
+			"method": "org.rdk.RDKShell.1.moveToBack",
+			"params": { "client": "ResidentApp"}
+			}`;
+		
+			ZorderChange.send(MoveToBackcmd);			
+			
+		}
+		};
+		
+	
+		SetVisibility.open("POST", "http://127.0.0.1:9998/jsonrpc");
+		SetVisibility.setRequestHeader("Accept", "application/json");
+		SetVisibility.setRequestHeader("Content-Type", "application/json");
+	
+		let SetVisibilityCmd = `{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"method": "org.rdk.RDKShell.1.setVisibility",
+		"params": { "client": "SmartScreen" , "visible": true }
+		}`;
+
+		SetVisibility.send(SetVisibilityCmd);
+		
         backAction = navigation.ToStart;
+
     },
     ToSuccess: () => {
         setupOptions.currentScreen = "success";
+        document.body.style.setProperty('background-color', '#1A2731', 'important');
         document.getElementById("root").style.display = "none"; // SS SDK container
         document.getElementById("ss_sdk_oobe").style.display = "block"; // OOBE container
         document.getElementById("screen_default_0").style.display = "none";
@@ -104,27 +244,207 @@ var navigation = {
         document.getElementById("screen_default_error").style.display = "none";
         document.getElementById("bt_back").style.display = "none";
         document.getElementById("bt_done").style.display = "block";
+        document.getElementById("bt_skip").style.display = "none";
+		
+		ClrElementFocus('success');
+		BtnElements =[document.getElementById("bt_done")];
+		
+		var SetVisibility = new XMLHttpRequest();
+	
+		SetVisibility.onreadystatechange = function() {
+		if (this.readyState == 4 && this.status == 200) {
+		
+			var ZorderChange = new XMLHttpRequest();
+
+			ZorderChange.onreadystatechange = function() {
+			if (this.readyState == 4 && this.status == 200) {
+		
+				var FocusReq = new XMLHttpRequest();
+				FocusReq.open("POST", "http://127.0.0.1:9998/jsonrpc");
+				FocusReq.setRequestHeader("Accept", "application/json");
+				FocusReq.setRequestHeader("Content-Type", "application/json");
+		
+				let Focuscmd = `{
+				"jsonrpc": "2.0",
+				"id": 3,
+				"method": "org.rdk.RDKShell.1.setFocus",
+				"params": { "client": "SmartScreen"}
+				}`;
+			
+				FocusReq.send(Focuscmd);
+			}
+			};
+			
+			ZorderChange.open("POST", "http://127.0.0.1:9998/jsonrpc");
+			ZorderChange.setRequestHeader("Accept", "application/json");
+			ZorderChange.setRequestHeader("Content-Type", "application/json");
+		
+			let MoveToBackcmd = `{
+			"jsonrpc": "2.0",
+			"id": 3,
+			"method": "org.rdk.RDKShell.1.moveToBack",
+			"params": { "client": "ResidentApp"}
+			}`;
+		
+			ZorderChange.send(MoveToBackcmd);			
+			
+		}
+		};
+		
+	
+		SetVisibility.open("POST", "http://127.0.0.1:9998/jsonrpc");
+		SetVisibility.setRequestHeader("Accept", "application/json");
+		SetVisibility.setRequestHeader("Content-Type", "application/json");
+	
+		let SetVisibilityCmd = `{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"method": "org.rdk.RDKShell.1.setVisibility",
+		"params": { "client": "SmartScreen" , "visible": true }
+		}`;
+
+		SetVisibility.send(SetVisibilityCmd);
+		
     },
     ToError: (errorType) => {
         setupOptions.currentScreen = "error";
+        document.body.style.setProperty('background-color', '#1A2731', 'important');
         document.getElementById("root").style.display = "none"; // SS SDK container
         document.getElementById("ss_sdk_oobe").style.display = "block"; // OOBE container
         document.getElementById("screen_default_0").style.display = "none";
         document.getElementById("screen_default_cbl").style.display = "none";
         document.getElementById("screen_default_success").style.display = "none";
         document.getElementById("screen_default_error").style.display = "block";
-        document.getElementById("bt_back").style.display = "block";
-        document.getElementById("bt_done").style.display = "none";
+        document.getElementById("bt_back").style.display = "none";
+        document.getElementById("bt_done").style.display = "block";
+        document.getElementById("bt_skip").style.display = "none";
+		
+		ClrElementFocus('error');
+		BtnElements =[document.getElementById("bt_done")];
 
         // display error message
         document.getElementById("screen_default_error_text").innerHTML = textStrings.genericError;
 
+		var SetVisibility = new XMLHttpRequest();
+	
+		SetVisibility.onreadystatechange = function() {
+		if (this.readyState == 4 && this.status == 200) {
+		
+			var ZorderChange = new XMLHttpRequest();
+
+			ZorderChange.onreadystatechange = function() {
+			if (this.readyState == 4 && this.status == 200) {
+		
+				var FocusReq = new XMLHttpRequest();
+				FocusReq.open("POST", "http://127.0.0.1:9998/jsonrpc");
+				FocusReq.setRequestHeader("Accept", "application/json");
+				FocusReq.setRequestHeader("Content-Type", "application/json");
+		
+				let Focuscmd = `{
+				"jsonrpc": "2.0",
+				"id": 3,
+				"method": "org.rdk.RDKShell.1.setFocus",
+				"params": { "client": "SmartScreen"}
+				}`;
+			
+				FocusReq.send(Focuscmd);
+			}
+			};
+			
+			ZorderChange.open("POST", "http://127.0.0.1:9998/jsonrpc");
+			ZorderChange.setRequestHeader("Accept", "application/json");
+			ZorderChange.setRequestHeader("Content-Type", "application/json");
+		
+			let MoveToBackcmd = `{
+			"jsonrpc": "2.0",
+			"id": 3,
+			"method": "org.rdk.RDKShell.1.moveToBack",
+			"params": { "client": "ResidentApp"}
+			}`;
+		
+			ZorderChange.send(MoveToBackcmd);			
+			
+		}
+		};
+		
+	
+		SetVisibility.open("POST", "http://127.0.0.1:9998/jsonrpc");
+		SetVisibility.setRequestHeader("Accept", "application/json");
+		SetVisibility.setRequestHeader("Content-Type", "application/json");
+	
+		let SetVisibilityCmd = `{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"method": "org.rdk.RDKShell.1.setVisibility",
+		"params": { "client": "SmartScreen" , "visible": true }
+		}`;
+
+		SetVisibility.send(SetVisibilityCmd);
+
         backAction = navigation.ToStart;
+
     },
     ToDone: () => {
         setupOptions.currentScreen = "done";
+        document.body.style.setProperty('background-color', 'transparent');
         document.getElementById("root").style.display = "block"; // SS SDK container
         document.getElementById("ss_sdk_oobe").style.display = "none"; // OOBE container
+
+	var SetVisibility = new XMLHttpRequest();
+	
+	SetVisibility.onreadystatechange = function() {
+    if (this.readyState == 4 && this.status == 200) {
+		
+		var ZorderChange = new XMLHttpRequest();
+
+		ZorderChange.onreadystatechange = function() {
+		if (this.readyState == 4 && this.status == 200) {
+		
+			var FocusReq = new XMLHttpRequest();
+			FocusReq.open("POST", "http://127.0.0.1:9998/jsonrpc");
+			FocusReq.setRequestHeader("Accept", "application/json");
+			FocusReq.setRequestHeader("Content-Type", "application/json");
+		
+			let Focuscmd = `{
+			"jsonrpc": "2.0",
+			"id": 3,
+			"method": "org.rdk.RDKShell.1.setFocus",
+			"params": { "client": "ResidentApp"}
+			}`;
+			
+			FocusReq.send(Focuscmd);
+		}
+		};
+		
+		ZorderChange.open("POST", "http://127.0.0.1:9998/jsonrpc");
+		ZorderChange.setRequestHeader("Accept", "application/json");
+		ZorderChange.setRequestHeader("Content-Type", "application/json");
+		
+		let MoveToFrontcmd = `{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"method": "org.rdk.RDKShell.1.moveToFront",
+		"params": { "client": "ResidentApp"}
+		}`;
+		
+		ZorderChange.send(MoveToFrontcmd);
+				
+    }
+	};
+	
+	SetVisibility.open("POST", "http://127.0.0.1:9998/jsonrpc");
+	SetVisibility.setRequestHeader("Accept", "application/json");
+	SetVisibility.setRequestHeader("Content-Type", "application/json");
+
+	let SetVisibilityCmd = `{
+	"jsonrpc": "2.0",
+	"id": 3,
+	"method": "org.rdk.RDKShell.1.setVisibility",
+	"params": { "client": "SmartScreen" , "visible": false }
+	}`;
+
+	SetVisibility.send(SetVisibilityCmd);
+		
     }
 }
 
@@ -154,6 +474,59 @@ var cblCodeAcquired = (cblData) => {
     }
 }
 
+//clear focus
+var ClrElementFocus = (currentpage) => {
+
+if(currentpage !== prvspage)
+{
+	TabIndexCounter = 0;
+	if(document.activeElement)
+	{
+		document.activeElement.blur();
+		LstPrsdKey = '';
+	}
+
+	prvspage = currentpage;
+}
+}
+
+//key events
+document.addEventListener('keydown', (event) => {
+
+  var name = event.key;
+  if (name === 'ArrowDown') {
+	  if(LstPrsdKey === 'ArrowUp')
+	  {
+		  TabIndexCounter = TabIndexCounter+1;
+	  }
+	  if(TabIndexCounter < BtnElements.length){
+		  BtnElements[TabIndexCounter].focus();
+		  TabIndexCounter = TabIndexCounter+1;
+		  LstPrsdKey = 'ArrowDown';
+	  }
+  }
+  else if(name === 'ArrowUp') {
+  
+	  if(LstPrsdKey === 'ArrowDown')
+	  {
+		  TabIndexCounter = TabIndexCounter-1;
+	  }
+	  
+		if(TabIndexCounter != 0){	
+			TabIndexCounter = TabIndexCounter-1;
+			BtnElements[TabIndexCounter].focus();
+			LstPrsdKey = 'ArrowUp';
+		}
+		else if(document.activeElement)
+		{
+			document.activeElement.blur();
+			LstPrsdKey = '';
+		}
+	  
+	  //}
+  }
+}, false);
+
 // bind events
 document.getElementById("bt_sign_in").addEventListener("click", () => {
     navigation.ToCBL();
@@ -166,3 +539,8 @@ document.getElementById("bt_back").addEventListener("click", () => {
 document.getElementById("bt_done").addEventListener("click", () => {
     navigation.ToDone();
 });
+
+document.getElementById("bt_skip").addEventListener("click", () => {
+    LoginSkip = 1;
+    navigation.ToDone();
+});
diff --git a/ss-sdk.html b/ss-sdk.html
index 8e8979c..1c46fc0 100755
--- a/ss-sdk.html
+++ b/ss-sdk.html
@@ -23,7 +23,6 @@
                   <button type="button" class="btn btn-secondary float-end" id="bt_done">Done</button>
             </div>
 
-
             <div class="body-content">
                <div class="text-center">
                   <img class="img-fluid" id="amazon_alexa_logo" src="img/Amz_Alexa_Logo_RGB_WHITE.png" alt="Alexa logo">
@@ -79,6 +78,11 @@
             </div>
 
          </div>
+	
+        <div class="text-center">
+	   <button type="button" class="btn btn-secondary float-end" id="bt_skip">Skip</button>
+	</div>
+
       </div>
 
    </div>
-- 
2.17.1

